python/ray/data/_internal/block_batching/iter_batches.py

```diff
@@ -1,47 +1,247 @@
+
+        # Step 3: Batch and shuffle the resolved blocks.
+        batch_iter = blocks_to_batches(
+            block_iter=block_iter,
+            stats=stats,
+            batch_size=batch_size,
+            drop_last=drop_last,
+            shuffle_buffer_min_size=shuffle_buffer_min_size,
+            shuffle_seed=shuffle_seed,
+            ensure_copy=ensure_copy,
+        )
+
+        # Step 4: Use a threadpool for formatting and collation.
+        batch_iter = _format_in_threadpool(
+            batch_iter,
+            stats=stats,
+            batch_format=batch_format,
+            collate_fn=collate_fn,
+            num_threadpool_workers=prefetch_batches,
+        )
+
+        # Step 5: Restore original order.
+        batch_iter: Iterator[Batch] = restore_original_order(batch_iter)
+
+        yield from extract_data_from_batch(batch_iter)
+
+    # Run everything in a separate thread to not block the main thread when waiting
+    # for streaming results.
+    async_batch_iter = make_async_gen(block_refs, fn=_async_iter_batches, num_workers=1)
+
+    while True:
+        with stats.iter_total_blocked_s.timer() if stats else nullcontext():
+            try:
+                next_batch = next(async_batch_iter)
+            except StopIteration:
+                break
+        with stats.iter_user_s.timer() if stats else nullcontext():
+            yield next_batch
+
+
+def _format_in_threadpool(
+    batch_iter: Iterator[Batch],
+    stats: DatasetStats,
+    batch_format: str = "default",
+    collate_fn: Optional[Callable[[DataBatch], Any]] = None,
+    num_threadpool_workers: int = 0,
+) -> Iterator[Batch]:
+    """Executes the batching, formatting, and collation logic in a threadpool.
+
+    Args:
+        logical_batch_iterator: An iterator over logical batches.
+        stats: DatasetStats object to record timing and other statistics.
+        batch_format: The format in which to return each batch.
+            Specify "default" to use the current block format (promoting
+            Arrow to pandas automatically), "pandas" to
+            select ``pandas.DataFrame`` or "pyarrow" to select
+            ``pyarrow.Table``. Default is "default".
+        collate_fn: A function to apply to each data batch before returning it.
+        num_threadpool_workers: The number of threads to use in the threadpool.
+    """
+
+    def threadpool_computations(
+        batch_iter: Iterator[Batch],
+    ) -> Iterator[Batch]:
+        # Step 4a: Format the batches.
+        formatted_batch_iter = format_batches(
+            batch_iter, batch_format=batch_format, stats=stats
+        )
+
+        # Step 4b: Apply the collate function if applicable.
+        if collate_fn is not None:
+            formatted_batch_iter = collate(
+                formatted_batch_iter, collate_fn=collate_fn, stats=stats
+            )
+        yield from formatted_batch_iter
+
+    if num_threadpool_workers > 0:
```

Comment:
Let's always use the thread pool?

Reply:
I don't this would be the correct behavior right? With a threadpool size of 1, we are prefetching 1 batch.

Reply:
Got it. I'm a little bit wondering if we should make the prefetch default 1, but I guess that could double memory usage, is that right? On the other hand, bad perf by default is arguably worse.

tf.data seems to auto-tune their prefetch by default.

Reply:
Yeah we can make 1 the default, but you're right, this will double the memory usage.

Agree, auto-tuning is the ideal that we want.